File: services/node-services/src/services.ts
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate e2e tests,
// which are released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/e2e/blob/main/LICENSE

import {
  ObjectContext,
  RestateEndpoint,
  Service,
  ServiceDefintion,
  VirtualObject,
  VirtualObjectDefintion,
} from "@restatedev/restate-sdk";

export type IComponent = {
  fqdn: string;
  binder: (endpoint: RestateEndpoint) => void;
};

export class ComponenetRegistery {
  constructor(readonly components: Map<string, IComponent> = new Map()) {}

  add(c: IComponent) {
    this.components.set(c.fqdn, c);
  }

  addObject(o: VirtualObjectDefintion<string, unknown>) {
    this.add({
      fqdn: o.path,
      binder: (b) => b.object(o),
    });
  }

  addService(
    s: ServiceDefintion<
      "Coordinator",
      Service<{
        sleep: (ctx: ObjectContext, request: number) => Promise<void>;
        manyTimers: (ctx: ObjectContext, request: number[]) => Promise<void>;
        proxy: (ctx: ObjectContext) => Promise<string>;
        complex: (
          ctx: ObjectContext,
          request: { sleepDurationMillis: number; requestValue: string }
        ) => Promise<string>;
        timeout: (ctx: ObjectContext, millis: number) => Promise<boolean>;
        invokeSequentially: () => never;
      }>
    >
  ) {
    this.add({
      fqdn: s.path,
      binder: (b) => b.service(s),
    });
  }

  register(fqdns: Set<string>, e: RestateEndpoint) {
    fqdns.forEach((fqdn) => {
      const c = this.components.get(fqdn);
      if (!c) {
        throw new Error(
          `unknown fqdn ${fqdn}. Did you rememeber to import the test at app.ts?`
        );
      }
      c.binder(e);
    });
  }
}

export const REGISTRY = new ComponenetRegistery();

File: services/go-services/services/services.go
package services

import (
	"log"

	"github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
)

var REGISTRY = Registry{components: map[string]Component{}}

type Registry struct {
	components map[string]Component
}

type Component struct {
	Fqdn   string
	Binder func(endpoint *server.Restate)
}

func (r *Registry) Add(c Component) {
	r.components[c.Fqdn] = c
}

func (r *Registry) AddRouter(name string, router restate.Router) {
	r.Add(Component{
		Fqdn:   name,
		Binder: func(e *server.Restate) { e.Bind(name, router) },
	})
}

func (r *Registry) Register(fqdns map[string]struct{}, e *server.Restate) {
	for fqdn := range fqdns {
		c, ok := r.components[fqdn]
		if !ok {
			log.Fatalf("unknown fqdn %s. Did you remember to import the test at app.ts?", fqdn)
		}
		c.Binder(e)
	}
}
